service/history/shard/controller_impl.go

```diff
@@ -219,6 +219,9 @@ func (c *ControllerImpl) getOrCreateShardContext(shardID int32) (*ContextImpl, e
 		if shard.isValid() {
 			return shard, nil
 		}
+
+		shard.Unload()
+		delete(c.historyShards, shardID)
```

Comment:
>as I said in the other PR, this doesn't make sense here. if the shard is in state stopping/stopped, unload does nothing

yes, but unloading again does no harm. 
one more thing, "IsValid" function now takes shutdown channel into consideration, so a shard is not in a valid state if shutdown chan is called.

Reply:
>delete does nothing since right below we just add it to the map

```go
if shard, ok := c.historyShards[shardID]; ok {
	if shard.isValid() {
		return shard, nil
	}

	shard.Unload()
	delete(c.historyShards, shardID) // <- delete removes a invalid shard 
}
```

Reply:
>if you intend for closing the shutdown channel to stop the shard context, then it shouldn't wait for someone to call getOrCreateShardContext to do that, it should do it proactively (maybe with another goroutine that just blocks on that)

i actually agrees with above, but ^ requires making the shard state management clear. i would consider <- as an issue to be solved later